index/segment/mem/simple_terms_dict.go

```diff
@@ -0,0 +1,213 @@
+		return nil, nil
+	}
+
+	// get postingList(s) for the given value.
+	lists, err := fieldValues.fetchLists(fieldValueFilter, opts.isRegexp)
+	if err != nil {
+		return nil, err
+	}
+
+	// union all the postingsList(s)
+	unionedList := t.opts.PostingsListPool().Get()
+	for _, ids := range lists {
+		unionedList.Union(ids)
+	}
+
+	return unionedList, nil
+}
+
+func (t *simpleTermsDictionary) getOrAddFieldName(fieldName string) *fieldValuesMap {
+	// cheap read lock to see if it already exists
+	t.fields.RLock()
+	fieldValues, ok := t.fields.nameToValuesMap[fieldName]
+	t.fields.RUnlock()
+	if ok {
+		return fieldValues
+	}
+
+	// acquire write lock and create
+	t.fields.Lock()
+	fieldValues, ok = t.fields.nameToValuesMap[fieldName]
+
+	// check if it's been created since we last acquired the lock
+	if ok {
+		t.fields.Unlock()
+		return fieldValues
+	}
+
+	fieldValues = newFieldValuesMap(t.opts)
+	t.fields.nameToValuesMap[fieldName] = fieldValues
+	t.fields.Unlock()
+	return fieldValues
+}
+
+type fieldValuesMap struct {
+	sync.RWMutex
+
+	opts Options
+
+	// fieldValue -> postingsList
+	values map[string]segment.PostingsList
+	// TODO: as noted in https://github.com/m3db/m3ninx/issues/11, evalute impact of using
+	// a custom hash map where we can avoid using string keys, both to save allocs and
+	// help perf.
+}
+
+func newFieldValuesMap(opts Options) *fieldValuesMap {
+	return &fieldValuesMap{
+		opts:   opts,
+		values: make(map[string]segment.PostingsList),
+	}
+}
+
+func (f *fieldValuesMap) addDocIDForValue(value string, i segment.DocID) error {
+	// try read lock to see if we already have a postingsList for the given value.
+	f.RLock()
+	pid, ok := f.values[value]
+	f.RUnlock()
+
+	// we have a postingsList, mark the docID and move on.
+	if ok {
+		return pid.Insert(i)
+	}
+
+	// postingsList doesn't exist, time to acquire write lock
+	f.Lock()
+	pid, ok = f.values[value]
+
+	// check if it's been created since we released lock
+	if ok {
+		f.Unlock()
+		return pid.Insert(i)
+	}
+
+	// create new posting id for the term, and insert into fieldValues
+	pid = f.opts.PostingsListPool().Get()
+	f.values[value] = pid
+	f.Unlock()
+	return pid.Insert(i)
+}
+
+// TODO: consider returning an iterator here, this would require some kind of ordering semantics
+// on the underlying map tho.
+func (f *fieldValuesMap) fetchLists(valueFilter []byte, regexp bool) ([]segment.PostingsList, error) {
+	// special case when we're looking for an exact match
+	if !regexp {
+		return f.fetchExact(valueFilter), nil
+	}
+
+	// otherwise, we have to iterate over all known values
+	pred, err := newRegexPredicate(valueFilter)
+	if err != nil {
+		return nil, err
+	}
+
+	f.RLock()
+	initLength := int(regexpPostingsListInitFactor * float64(len(f.values)))
+	postingsLists := make([]segment.PostingsList, 0, initLength)
+	for value, list := range f.values {
+		// TODO: evaluate lock contention caused by holding on to the read lock
+		// while evaluating this predicate.
```

Comment:
You can do what we do in the M3DB shard lookup, keep both a list and a map, then iterate over the list in a batched fashion to avoid holding lock for an entire scan and use lock in batches of work.

Reply:
Yep, for sure. The hope is the map backed segment never gets large enough for this to be an issue, will address if it does.

Reply:
done